File: robotics-codes-from-scratch-master/python/IK_nullspace.py
'''
Inverse kinematics example

Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
Written by Sylvain Calinon <https://calinon.ch>

This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
License: MIT
'''

import numpy as np
import matplotlib.pyplot as plt


# Logarithmic map for R^2 x S^1 manifold
def logmap(f, f0):
	position_error = f[:2,:] - f0[:2,:]
	orientation_error = np.imag(np.log(np.exp(f0[-1,:]*1j).conj().T * np.exp(f[-1,:]*1j).T)).conj()
	diff = np.vstack([position_error, orientation_error])
	return diff

# Forward kinematics for end-effector (in robot coordinate system)
def fkin(x, param):
	L = np.tril(np.ones([param.nbVarX, param.nbVarX]))
	f = np.vstack([
		param.l @ np.cos(L @ x),
		param.l @ np.sin(L @ x),
		np.mod(np.sum(x,0)+np.pi, 2*np.pi) - np.pi
	]) # f1,f2,f3, where f3 is the orientation (single Euler angle for planar robot)
	return f

# Forward kinematics for all joints (in robot coordinate system)
def fkin0(x, param): 
	L = np.tril(np.ones([param.nbVarX, param.nbVarX]))
	f = np.vstack([
		L @ np.diag(param.l) @ np.cos(L @ x),
		L @ np.diag(param.l) @ np.sin(L @ x)
	])
	f = np.hstack([np.zeros([2,1]), f])
	return f

# Jacobian with analytical computation (for single time step)
def Jkin(x, param):
	L = np.tril(np.ones([param.nbVarX, param.nbVarX]))
	J = np.vstack([
		-np.sin(L @ x).T @ np.diag(param.l) @ L,
		 np.cos(L @ x).T @ np.diag(param.l) @ L,
		 np.ones([1,param.nbVarX])
	])
	return J

## Parameters
# ===============================

param = lambda: None # Lazy way to define an empty class in python
param.dt = 1e-2 # Time step length
param.nbData = 50 # Number of datapoints
param.nbVarX = 3 # State space dimension (x1,x2,x3)
param.nbVarU = 3 # Control space dimension (dx1,dx2,dx3)
param.nbVarF = 3 # Objective function dimension (position and orientation of the end-effector)
param.l = [2, 2, 1] # Robot links lengths

fh = np.array([4, 1, -np.pi/2]) #Desired target for the end-effector (position and orientation)
x = -np.ones(param.nbVarX) * np.pi / param.nbVarX # Initial robot pose
x[0] = x[0] + np.pi 

## Inverse kinematics (IK)
# ===============================

plt.scatter(fh[0], fh[1], color='r', marker='.', s=10**2) #Plot target
for t in range(param.nbData):
	f = fkin(x, param) # Forward kinematics (for end-effector)
	dfp = (fh[:2] - f[:2,0]) * 20 # Position correction
	dfo = (fh[2] - f[2:,0]) * 20 # Orientation correction
	
	J = Jkin(x, param) # Jacobian (for end-effector)
	Jp = J[:2,:] # Jacobian for position 
	Jo = J[2:,:] # Jacobian for orientation
	pinvJp = np.linalg.inv(Jp.T @ Jp + np.eye(param.nbVarX) * 1e-4) @ Jp.T # Damped pseudoinverse
	Np = np.eye(param.nbVarX) - pinvJp @ Jp # Nullspace projection operator
	up = pinvJp @ dfp # Command for position tracking
	JoNp = Jo @ Np
	pinvJoNp = JoNp.T @ np.linalg.inv(JoNp @ JoNp.T + 1e-1) # Damped pseudoinverse
	uo = pinvJoNp @ (dfo - Jo @ up) # Command for orientation tracking (with position tracking prioritized)
	
	x += (up + Np @ uo) * param.dt # Update state 
	f_rob = fkin0(x, param) # Forward kinematics (for all articulations, including end-effector)
	plt.plot(f_rob[0,:], f_rob[1,:], color=str(1-t/param.nbData), linewidth=1) # Plot robot

plt.axis('off')
plt.axis('equal')
plt.show()

File: robotics-codes-from-scratch-master/matlab/IK_nullspace.m
%% Inverse kinematics with nullspace projection operator and line search
%% (position and orientation tracking as primary or secondary tasks)
%% 
%% Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
%% Written by Teguh Lembono <teguh.lembono@idiap.ch> and 
%% Sylvain Calinon <https://calinon.ch>
%% 
%% This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
%% License: MIT

function IK_nullspace

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%param.dt = 1E-2; %Time step
param.nbIter = 100; %Number of iterations
param.Kp = 1E0; %Position tracking gain
param.Ko = 1E0; %Orientation tracking gain
param.fh = [4; 1; -pi/2]; %Target

param.mu_p = 1E-4; %regularization for position
param.mu_o = 1E-4; %regularization for orientation
param.alpha_fac = 0.5;

%% Robot parameters 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbVarX = 3; %Number of articulations
param.l = [2; 2; 1]; %Link lengths

%% Initialization
x = [3*pi/4; -pi/2; -pi/4]; %Initial pose
h = figure('position',[10,10,800,800],'color',[1,1,1]); hold on; axis off;
plot(param.fh(1,:), param.fh(2,:), 'r.','markersize',30); %Plot target


%% IK with nullspace
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
for t=1:param.nbIter
	%Calculate forward kinematics	
	f = fkin(x, param);

	%Jacobians and pseudoinverses
	J = Jkin(x, param);
	Jp = J(1:2,:); %Jacobian for position 
	Jo = J(3,:); %Jacobian for orientation
	pinvJp = (Jp' * Jp + eye(param.nbVarX) * param.mu_p) \ Jp'; %Damped pseudoinverse
	Np = eye(param.nbVarX) - pinvJp * Jp; %Nullspace projection operators

	%Error corrections
	dfp = param.Kp * (param.fh(1:2) - f(1:2));
	dfo = param.Ko * (param.fh(3) - f(3));

	%Prioritized tracking task 
	%(position tracking as primary task, orientation tracking as secondary task)

	%1. Determine the first step size
	alpha1 = 1;
	cost10 = (param.fh(1:2) - f(1:2))' * (param.fh(1:2) - f(1:2)); %current cost
	cost1 = 5 * cost10; %initialize with something big
	dxp = pinvJp * dfp; %search direction for first cost
	while cost1 > cost10 + 1e-4 %1e-4 is the tolerance
		dxp_try = alpha1 * dxp;
		x_next = x + dxp_try;
		f = fkin(x_next, param);
		cost1 = (param.fh(1:2) - f(1:2))' * (param.fh(1:2) - f(1:2));
		alpha1 = alpha1 * param.alpha_fac;
	end
	dxp = dxp_try;
	
	x_next = x + dxp;
	f = fkin(x_next, param);
  
	
	%2. Determine the second step size
	alpha2 = 1;
	cost10 = (param.fh(1:2) - f(1:2))' * (param.fh(1:2) - f(1:2)); %current cost1
	cost20 = (param.fh(3) - f(3))' * (param.fh(3) - f(3)); %current cost2

	cost1 = 5 * cost10 + 1e1; %initialize with something big
	cost2 = 5 * cost20; %initialize with something big

	JoNp = Jo * Np;
	pinvJoNp = JoNp' / (JoNp * JoNp' + eye(1) * param.mu_o);
	dxo = Np * pinvJoNp * (dfo - Jo * dxp); %search direction for second cost
  
	%Decrease alpha2 until the cost 1 and cost2 are lower or the same
	%(with some tolerance)
	while (cost1 > cost10 + 1e-8) || (cost2 > cost20 + 1e-4) %1e-8 is the tolerance
		dxo_try = alpha2 * dxo;
		x_next = x + dxp + dxo_try;
		f = fkin(x_next, param);
		cost1 = (param.fh(1:2) - f(1:2))' * (param.fh(1:2) - f(1:2));
		cost2 = (param.fh(3) - f(3))' * (param.fh(3) - f(3));
		alpha2 = alpha2 * param.alpha_fac;
	end
	dxo = dxo_try;
	
	dx = dxp + dxo;
	x = x + dx ; %Update pose

	fprintf('Cost_p: %f, Cost_o: %f, alpha1: %f, alpha2:%f\n', ...
	(param.fh(1:2) - f(1:2))' * (param.fh(1:2) - f(1:2)), ...
	(param.fh(3) - f(3)) * (param.fh(3) - f(3)), alpha1, alpha2);
	
	%Plot the robot
	fs = fkin0(x, param);
	%plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.8 .8 .8]);
	plot(fs(1,:), fs(2,:), '-','linewidth',4,'color',ones(1,3)*(1-t/param.nbIter)); 

end
%Plot the final configuration
fs = fkin0(x, param);
plot(fs(1,:), fs(2,:), '-','linewidth',4,'color',[.4 .4 .4]); 
axis equal; 
waitfor(h);
end


%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for all robot articulations (in robot coordinate system)
function f = fkin0(x, param)
	L = tril(ones(size(x,1)));
	f = [L * diag(param.l) * cos(L * x), ...
	     L * diag(param.l) * sin(L * x)]'; 
	f = [zeros(2,1), f];
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for end-effector 
function f = fkin(x, param)
	L = tril(ones(size(x,1)));
	f = [param.l' * cos(L * x); ...
	     param.l' * sin(L * x); ...
	     mod(sum(x,1)+pi, 2*pi) - pi]; %f1,f2,f3, where f3 is the orientation (single Euler angle for planar robot)
end

%%%%%%%%%%%%%%%%%%%%%%
%Jacobian with numerical computation
function J = Jkin(x, l)
	e = 1E-6;
	D = size(x,1);
	
%	%Slow for-loop computation
%	J = zeros(3,D);
%	for n=1:D
%		xtmp = x;
%		xtmp(n) = xtmp(n) + e;
%		ftmp = fkin(xtmp, l); %Forward kinematics 
%		f = fkin(x, l); %Forward kinematics 
%		J(:,n) = (ftmp - f) / e; %Jacobian for end-effector
%	end
	
	%Fast matrix computation
	X = repmat(x, [1, D]);
	F1 = fkin(X, l);
	F2 = fkin(X + eye(D) * e, l);
	J = (F2 - F1) / e;
end
